Fix loading of ET_DYN type of shared objects

nealef opened the pull request:
* build-globals.sh
  - Only look in the nuttx for external symbols used when loading dynamic shared objects

* include/elf64.h
  - Correct the type of fields in the Elf64_Phdr structure

* libs/libc/dlfcn/lib_dlclose.c
  - Distinguish between ET_DYN and other objects as the former has both text and data in a single allocation to reserve GOT offsets

* libs/libc/dlfcn/lib_dlopen.c
  - Code formatting

* libs/libc/modlib/modlib_bind.c
  - Distinguish between relocation entry sizes by section type
  - Handle RELA style relocations

* libs/libc/modlib/modlib_globals.S
  - Formatting fixes
  - Symbols should not be weak - they exist or they don't

* include/nuttx/lib/modlib.h
  - Add an inidcator to module_s to distinguish between ET_DYN and other

* libs/libc/modlib/modlib_load.c
  - ET_DYN objects need to keep the relative displacement between the text and data sections due to GOT references from the former to the latter. This also implies that linking may require modification from the default for the shared objects being produced. For example, default alignment may mean nearly 64K of wasted space.

* libs/libc/modlib/modlib_unload.c sched/module/mod_rmmod.c
  - Distingusih between freeing of ET_DYN storage and other as the former is a single allocation.

* libs/libc/modlib/mod_insmod.c
  - Cater for ET_DYN objects having init and pre-init sections

## Summary

An ET_DYN type of shared objects relies on its text section and data sections remaining in relative position as there are references from the text section to GOT offsets in the data section. Therefore you cannot just arbitrarily load text and data separately and hope the offsets are not changed. 

## Impact

ET_DYN objects will load, execute, and unload correctly.

## Testing

Added test to [sotest](nuttx-apps/examples/sotest)
https://github.com/apache/nuttx-apps/pull/2026

xiaoxiang781216 commented:
please rebase to the last master to fix the conflict.

acassis commented:
@masayuki2009 please check this PR to confirm it works for you too, I saw you reverted the https://github.com/apache/nuttx/pull/10471 so it is important to test early

pkarashchenko commented:
The PR needs rebase on the fresh master

masayuki2009 commented:
@nealef 
Please rebase this PR again to reflect https://github.com/apache/nuttx/pull/10474

masayuki2009 commented:
@nealef 

Please `git rebase -i HEAD~4` to fixup the commits into one commit.

```
commit 4c925d0a322e1f5427f3e63e7ebbbb711b8b2d10 (HEAD -> dynload-fix, origin/dynload-fix)
Author: Neale Ferguson <neale@sinenomine.net>
Date:   Wed Sep 6 13:26:15 2023 +1000

    Style fix and avoid a compiler warning

    * libs/libc/modlib/modlib_bind.c
      - Style
      - Avoid compiler warning for a situation that won't happen

    * libs/libc/sched/module/mod_rmmod.c
      - Style

commit c31999cd49ed2e79901d115e0cfd9ac267f6fd99
Author: Neale Ferguson <neale@sinenomine.net>
Date:   Wed Sep 6 12:49:43 2023 +1000

    Revert some style fixes as block comments aren't what the style guide implies

commit 3f9cf64fb448dbf09d249e2e737c180388b494b7
Author: Neale Ferguson <neale@sinenomine.net>
Date:   Wed Sep 6 12:28:36 2023 +1000

    Further style fixes

    * libs/libc/dlfcn/lib_dlclose.c
      libs/libc/modlib/modlib_bind.c
      libs/libc/modlib/modlib_load.c
      libs/libc/modlib/modlib_unload.c
      sched/module/mod_rmmod.c

commit 38222bfc2fdd5eaf147dc7e2acee4b076d9909a4
Author: Neale Ferguson <neale@sinenomine.net>
Date:   Wed Sep 6 12:03:09 2023 +1000

    Stle updates

    * libs/libc/dlfcn/lib_dlclose.c
      libs/libc/modlib/modlib_bind.c
      libs/libc/modlib/modlib_load.c
      libs/libc/modlib/modlib_unload.c
      sched/module/mod_rmmod.c
      - Fixes to conform to style guide:
        - Brace alignment
        - Indentations
        - Danlging white space
        - Comments
```

nealef commented:
WRT squashing - I will do so as soon as I win the battle with the "Check" phases of the CI. There's at least another to come from the current build.

nealef commented:
@masayuki2009 Found and built `nxstyle` and ran it against my stuff and now get passes for everything. If I have an outstanding change can I add/commit and then run the command you specified using `HEAD~5` or does it need pushing first?

xiaoxiang781216 commented:
@nealef you can simply run before pushing the change:
```
./tools/checkpatch.sh -g HEAD~7...HEAD
```

nealef commented:
About to squash with: `git rebase -i HEAD~8` which shows:
```
pick 66287b7931 Fix loading of ET_DYN type of shared objects
fixup 012bc030b2 * libs/libc/dlfcn/lib/lib_dlclose.c   - Fix freeing the textalloc area twice
fixup f47f0505c3 Add an empty modlib_global.S - used only when end-user builds and loads their own dynamic shared object
fixup 38222bfc2f Stle updates
fixup 3f9cf64fb4 Further style fixes
fixup c31999cd49 Revert some style fixes as block comments aren't what the style guide implies
fixup 4c925d0a32 Style fix and avoid a compiler warning
fixup a9c257a6a0 Final style fixes
```
Look okay?

acassis commented:
> About to squash with: `git rebase -i HEAD~8` which shows:
> 
> ```
> pick 66287b7931 Fix loading of ET_DYN type of shared objects
> fixup 012bc030b2 * libs/libc/dlfcn/lib/lib_dlclose.c   - Fix freeing the textalloc area twice
> fixup f47f0505c3 Add an empty modlib_global.S - used only when end-user builds and loads their own dynamic shared object
> fixup 38222bfc2f Stle updates
> fixup 3f9cf64fb4 Further style fixes
> fixup c31999cd49 Revert some style fixes as block comments aren't what the style guide implies
> fixup 4c925d0a32 Style fix and avoid a compiler warning
> fixup a9c257a6a0 Final style fixes
> ```
> 
> Look okay?

Hi @nealef after the squash you should see a single commit when you run "git log" (it is your 7 commits squashed as a single commit).

masayuki2009 commented:
@nealef 

>Look okay?

In my understanding, this PR would have 2 commits.

The first one is "Fix loading of ET_DYN type of shared objects."

The second would be "Style fix for modlib and dlfcn". 
To change the commit log, you need to use `reword`.

So the `git rebase -i HEAD~8` would be something like

```
pick 66287b7931 Fix loading of ET_DYN type of shared objects
fixup 012bc030b2 * libs/libc/dlfcn/lib/lib_dlclose.c   - Fix freeing the textalloc area twice
fixup f47f0505c3 Add an empty modlib_global.S - used only when end-user builds and loads their own dynamic shared object
reword 38222bfc2f Stle updates
fixup 3f9cf64fb4 Further style fixes
fixup c31999cd49 Revert some style fixes as block comments aren't what the style guide implies
fixup 4c925d0a32 Style fix and avoid a compiler warning
fixup a9c257a6a0 Final style fixes
```

nealef commented:
Too late I'm afraid. Let me know if I can fix things if they don't suit. I note macOS sim-_nn_ are having problems that I don't think are related to my fix. At least check now works without complaint.

xiaoxiang781216 commented on libs/libc/modlib/modlib_bind.c:
```diff
@@ -717,7 +752,11 @@ static int modlib_relocatedyn(FAR struct module_s *modp,
-                    *(FAR uintptr_t *)addr = (uintptr_t)ep;
+
+                    if (reldata.relrela[idx_rel] == 1)
+                        addr += rela->r_addend;
```
{}

xiaoxiang781216 commented on libs/libc/modlib/modlib_bind.c:
```diff
@@ -717,7 +752,11 @@ static int modlib_relocatedyn(FAR struct module_s *modp,
+                    if (reldata.relrela[idx_rel] == 1)
+                        addr += rela->r_addend;
+
+                    *(uintptr_t *)addr = (uintptr_t)ep;
```
add FAR and space

xiaoxiang781216 commented on boards/sim/sim/sim/scripts/Make.defs:
```diff
@@ -230,6 +230,8 @@ CMODULEFLAGS += -fno-stack-protector
 
 LDMODULEFLAGS = -r -e module_initialize --gc-sections
 LDMODULEFLAGS += -T $(call CONVERT_PATH,$(TOPDIR)/libs/libc/modlib/gnu-elf.ld)
+SHLDMODULEFLAGS = -Bsymbolic -G -Bdynamic --gc-sections
```
SHLDMODULEFLAGS -> SHMODULEFLAGS?

xiaoxiang781216 commented:
@nealef please squash the style change to original patch. And fix the style issue locally:
```
./tools/checkpatch.sh -g HEAD~3...HEAD
```

pkarashchenko commented on libs/libc/modlib/modlib_globals.S:
```diff
@@ -54,20 +49,20 @@
 	.align	ALIGN
 	GLOBAL	nglobals
-SYMBOL(nglobals):
+SYMBOL(nglobals):	
```
```suggestion
SYMBOL(nglobals):
```

pkarashchenko commented on libs/libc/modlib/modlib_globals.S:
```diff
@@ -54,20 +49,20 @@
 
 SYMBOL(globalNames):
-	SIZE globalNames
+	SIZE	globalNames 
```
```suggestion
	SIZE	globalNames
```

pkarashchenko commented on libs/libc/modlib/modlib_globals.S:
```diff
@@ -54,20 +49,20 @@
 	.data
 	.align	ALIGN
-	GLOBAL	globalNames
+	GLOBAL	globalNames 
```
```suggestion
	GLOBAL	globalNames
```

pkarashchenko reviewed (commented):
Could you please rebase on top of the latest master so we can get better CI feedback?

xiaoxiang781216 commented:
@nealef you need do the following action to resolve the conflict:
```
git remote add apache git@github.com:apache/nuttx.git
git fetch apache
git rebase apache/master
git push -f origin
```

nealef commented:
I think the last failure was due to the push to apps occurring after the push to here. That push has a fix for the setting of the linker flags for dynload.

pkarashchenko commented:
Please rebase on top of the latest master, so we can get a clear feedback from CI and MACOS CI has been fixed over the weekend

nealef commented:
I rebased this morning (UTC+10) and corrected a merge conflict it flagged. I also adopted the suggestions in the PR relating to build-global.sh.


-------- Original message --------
From: Petro Karashchenko ***@***.***>
Date: 18/9/23 18:30 (GMT+10:00)
To: apache/nuttx ***@***.***>
Cc: Neale Ferguson ***@***.***>, Mention ***@***.***>
Subject: Re: [apache/nuttx] Fix loading of ET_DYN type of shared objects (PR #10470)


Please rebase on top of the latest master, so we can get a clear feedback from CI and MACOS CI has been fixed over the weekend

—
Reply to this email directly, view it on GitHub<https://github.com/apache/nuttx/pull/10470#issuecomment-1722965905>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AACLN6WLWLVJSY2CPRHGIOLX3ABCNANCNFSM6AAAAAA4JVX5FQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

pkarashchenko commented:
Yeah, indeed, now the error comes not from `libuv`, but 
```
Configuration/Tool: sim/libcxxtest
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Building NuttX...
Error: libcxx/src/filesystem/directory_iterator.cpp:189:57: error: unqualified call to 'std::move' [-Werror,-Wunqualified-std-cast-call]
                                                __root_(move(other.__root_)),
                                                        ^
                                                        std::
Error: libcxx/src/filesystem/directory_iterator.cpp:190:58: error: unqualified call to 'std::move' [-Werror,-Wunqualified-std-cast-call]
                                                __entry_(move(other.__entry_)) {
                                                         ^
                                                         std::
Error: libcxx/src/filesystem/directory_iterator.cpp:272:17: error: unqualified call to 'std::move' [-Werror,-Wunqualified-std-cast-call]
    path root = move(__imp_->__root_);
                ^
                std::
Error: libcxx/src/filesystem/directory_iterator.cpp:306:25: error: unqualified call to 'std::move' [-Werror,-Wunqualified-std-cast-call]
  __imp_->__stack_.push(move(new_s));
                        ^
                        std::
Error: libcxx/src/filesystem/directory_iterator.cpp:360:17: error: unqualified call to 'std::move' [-Werror,-Wunqualified-std-cast-call]
    path root = move(stack.top().__root_);
                ^
                std::
Error: libcxx/src/filesystem/directory_iterator.cpp:394:29: error: unqualified call to 'std::move' [-Werror,-Wunqualified-std-cast-call]
      __imp_->__stack_.push(move(new_it));
                            ^
                            std::
Error: libcxx/src/filesystem/directory_iterator.cpp:405:21: error: unqualified call to 'std::move' [-Werror,-Wunqualified-std-cast-call]
      path at_ent = move(curr_it.__entry_.__p_);
                    ^
                    std::
7 errors generated.
make[1]: *** [directory_iterator.o] Error 1
make[1]: Target `libxx.a' not remade because of errors.
make: *** [libs/libxx/libxx.a] Error 2
make: Target `all' not remade because of errors.
/Users/runner/work/nuttx/nuttx/sources/nuttx/tools/testbuild.sh: line 370: /Users/runner/work/nuttx/nuttx/sources/nuttx/../nuttx/nuttx.manifest: No such file or directory
```

pkarashchenko commented on libs/libc/modlib/modlib_bind.c:
```diff
@@ -594,8 +600,8 @@ static int modlib_relocatedyn(FAR struct module_s *modp,
             break;
-          case DT_RELENT:
-            reldata.relentsz = dyn[i].d_un.d_val;
+          case DT_RELENT :
```
```suggestion
          case DT_RELENT:
```

pkarashchenko commented on libs/libc/modlib/modlib_bind.c:
```diff
@@ -609,6 +615,18 @@ static int modlib_relocatedyn(FAR struct module_s *modp,
           case DT_PLTRELSZ:
             reldata.relsz[I_PLT] = dyn[i].d_un.d_val;
             break;
+          case DT_PLTREL :
```
```suggestion
          case DT_PLTREL:
```

xiaoxiang781216 commented:
this problem is fixed by https://github.com/apache/nuttx/pull/10690, you need fetch and rebase to the new master.

nealef commented:
Now `duktape` is failing. Should I wait until fixed and rebase again?

xiaoxiang781216 commented:
> Now `duktape` is failing. Should I wait until fixed and rebase again?

both are fixed by: https://github.com/apache/nuttx/pull/10699 and https://github.com/apache/nuttx-apps/pull/2073.
